### PR TITLE
Implementation of full profile tests

### PR DESCRIPTION
**Pull request by juangon**

### CODE REVIEW

File: hollow-jar/full/pom.xml [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.awaitility</groupId>
+            <artifactId>awaitility</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.icegreen</groupId>
+            <artifactId>greenmail</artifactId>
```
Comment (Ladicek) [inline hollow-jar/full/pom.xml]: Should be `<scope>test</scope>`?

Comment (juangon) [reply hollow-jar/full/pom.xml]: I think it didn't compile when I tried it..

Comment (Ladicek) [reply hollow-jar/full/pom.xml]: I see you're creating GreenMail inside the servlet. I'd expect that to be set up by the test; the servlet shouldn't know anything about us mocking the SMTP server.

File: hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/MailServlet.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        if (operation.equals("send")) {
+
+            ServerSetup serverSetup = new ServerSetup(3026, "127.0.0.1", ServerSetup.PROTOCOL_SMTP);
+            greenMail = new GreenMail (serverSetup);
+            greenMail.start ();
+
+            Message message = new MimeMessage(mailSession);
+
+            try {
+                message.setRecipient(Message.RecipientType.TO, new InternetAddress("swarm-test-suite@test.com"));
```
Comment (Ladicek) [inline hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/MailServlet.java]: Should probably use `@example.com`.

File: hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/MessagingServlet.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+        if (operation.equals("sendQueue")) {
+
+            processResult.clear();
+
+            JMSContext context = connectionFactory.createContext();
+
+            context.createProducer().send(queue, "1");
+            resp.getWriter().print("OK");
+        } else if (operation.equals("sendTopic")) {
```
Comment (Ladicek) [inline hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/MessagingServlet.java]: Indentation in this `else` block.

File: hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/ProcessResult.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Named;
+
+@Named
+@ApplicationScoped
+public class ProcessResult {
+
+    private volatile List<Object> writtenItems = new ArrayList<Object>();
+
+    public synchronized List<Object> getWrittenItems() {
```
Comment (Ladicek) [inline hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/ProcessResult.java]: The way you're doing synchronization here is useless. Either all methods accessing `writtenItems` need to be `synchronized`, or none of them. Also, you're iterating on the content of the collection in other classes, and that would have to be synchronized as well. Do we really have multiple threads accessing the `ProcessResult`? If no, let's avoid synchronization altogether. If yes, maybe we could use something like `CopyOnWriteArrayList`.

The `writtenItems` field doesn't really have to be `volatile`.

Comment (juangon) [reply hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/ProcessResult.java]: Yes I forgot remove this part before commit sorry :-(

Comment (juangon) [reply hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/ProcessResult.java]: This is fixed when using await() so we avoid concurrent access.

File: hollow-jar/full/src/test/java/org/wildfly/swarm/ts/hollow/jar/full/FullHollowJarIT.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    public void testBatch() throws IOException {
+        String response = Request.Get("http://localhost:8080/batch?operation=start").execute().returnContent().asString();
+        assertThat(response).isEqualTo("1");
+
+        String resultsURL = "http://localhost:8080/batch?operation=results";
+
+        await().
+            atLeast(10, TimeUnit.MILLISECONDS).
+            atMost(20, TimeUnit.MILLISECONDS).
+        until(() -> {
```
Comment (Ladicek) [inline hollow-jar/full/src/test/java/org/wildfly/swarm/ts/hollow/jar/full/FullHollowJarIT.java]: Better do `await().untilAsserted(...)` and do the assert inside that.

File: hollow-jar/full/src/test/java/org/wildfly/swarm/ts/hollow/jar/full/FullHollowJarIT.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    }
+
+    @Test
+    public void testJaxWS() throws IOException {
+        String response = Request.Get("http://localhost:8080/TestWebService?wsdl").execute().returnContent().asString();
+
+        assertThat(response).isNotBlank();
+    }
+}
```
Comment (Ladicek) [inline hollow-jar/full/src/test/java/org/wildfly/swarm/ts/hollow/jar/full/FullHollowJarIT.java]: Where's a test for the resource adapter (`net.java.xadisk`)?

File: hollow-jar/full/src/main/webapp/index.jsp [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<%@ taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c" %>
+<html>
+<head>
+    <title>JSP + JSTL Test</title>
+</head>
+<body>
+    <c:out value="Hello from JSP + JSTL"/>
+</body>
+</html>
```
Comment (Ladicek) [inline hollow-jar/full/src/main/webapp/index.jsp]: Do we need this file?

File: hollow-jar/full/pom.xml [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                          <goal>start</goal>
+                        </goals>
+                        <configuration>
+                            <name>swarm</name>
+                            <workingDir>${project.build.directory}</workingDir>
+                            <arguments>
+                                <argument>${java.home}/bin/java</argument>
+                                <argument>-jar</argument>
+                                <argument>dependency/full-hollowswarm.jar</argument>
+                                <argument>dependency/xadisk.rar</argument>
```
Comment (Ladicek) [inline hollow-jar/full/pom.xml]: I'd like to have _one_ execution for everything. If we can't easily deploy the RAR with a hollow JAR, then that's a bug.

File: hollow-jar/full/src/test/java/org/wildfly/swarm/ts/hollow/jar/full/FullHollowJarIT.java (hunk 6 above)
Comment (juangon) [reply hollow-jar/full/src/test/java/org/wildfly/swarm/ts/hollow/jar/full/FullHollowJarIT.java]: If it deploys, it works. I think I could get the remote resource adapter resource (as we have two different java -jar processes) but I would need the port etc and make things a bit more complicated, probably without the actual need. In current resource adapter tests it just test it is deploying without errors.

File: hollow-jar/full/src/main/webapp/index.jsp (hunk 7 above)
Comment (juangon) [reply hollow-jar/full/src/main/webapp/index.jsp]: TooMuchCopyPasteException. Right

File: hollow-jar/full/pom.xml (hunk 8 above)
Comment (juangon) [reply hollow-jar/full/pom.xml]: We can, but we need to provide as a deployment (instead of a WAR), right?

File: hollow-jar/full/src/test/java/org/wildfly/swarm/ts/hollow/jar/full/FullHollowJarIT.java (hunk 6 above)
Comment (Ladicek) [reply hollow-jar/full/src/test/java/org/wildfly/swarm/ts/hollow/jar/full/FullHollowJarIT.java]: As I wrote above, we don't want to have two processes -- with that, we don't really test that one can deploy a resource adapter together with an actual application. Also doing some simple operation with the resource adapter shouldn't be too hard :-)

File: hollow-jar/full/pom.xml (hunk 8 above)
Comment (Ladicek) [reply hollow-jar/full/pom.xml]: Well we _need to_ be able to deploy a RAR together with an application. Just like with regular fat JAR.

**Comment by juangon** [comment]
@Ladicek pushed suggested changes.

**Comment by juangon** [comment]
Sorry, now it is. Forgot to commit a new file

File: hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/MailServlet.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+            Multipart multipart = new MimeMultipart();
+            MimeBodyPart text = new MimeBodyPart();
+
+            text.setText("FOO");
+            multipart.addBodyPart(text);
+            message.setContent(multipart);
+            Transport.send(message);
+            resp.getWriter().print("OK");
+        } catch (MessagingException e) {
+            e.printStackTrace();
```
Comment (Ladicek) [inline hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/MailServlet.java]: Just rethrow as `ServletException`.

File: hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/RemoteEJBServlet.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.wildfly.swarm.ts.hollow.jar.full.remoteejb.RemoteEJB;
+
+@WebServlet("/remoteejb")
+public class RemoteEJBServlet  extends HttpServlet {
+
+    @EJB (lookup="java:module/RemoteEJBImpl")
```
Comment (Ladicek) [inline hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/RemoteEJBServlet.java]: Are we sure this really gets a remote proxy?

Also extra space between `@EJB` and opening `(`.

Comment (Ladicek) [reply hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/RemoteEJBServlet.java]: Maybe should invoke the EJB remotely from the test code.

File: hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/ResourceAdapterServlet.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    private XADiskConnectionFactory xaConnectionFactory;
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        try {
+            if (xaConnectionFactory != null && xaConnectionFactory.getConnection() != null) {
+                resp.getWriter().print("OK");
+            }
+        } catch (ResourceException e) {
+            e.printStackTrace();
```
Comment (Ladicek) [inline hollow-jar/full/src/main/java/org/wildfly/swarm/ts/hollow/jar/full/ResourceAdapterServlet.java]: Just rethrow as `ServletException`.

File: hollow-jar/full/src/test/java/org/wildfly/swarm/ts/hollow/jar/full/FullHollowJarIT.java [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    @Before
+    public void setup() {
+        Awaitility.setDefaultPollInterval(10, TimeUnit.MILLISECONDS);
+    }
+
+    @BeforeClass
+    public static void setupClass() {
+        ServerSetup serverSetup = new ServerSetup(3026, "127.0.0.1", ServerSetup.PROTOCOL_SMTP);
+        greenMail = new GreenMail (serverSetup);
+        greenMail.start ();
```
Comment (Ladicek) [inline hollow-jar/full/src/test/java/org/wildfly/swarm/ts/hollow/jar/full/FullHollowJarIT.java]: Extra spaces before opening `(`.

File: hollow-jar/full/src/test/java/org/wildfly/swarm/ts/hollow/jar/full/FullHollowJarIT.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        String resultsURL = "http://localhost:8080/messaging?operation=results";
+        await().
+            atLeast(10, TimeUnit.MILLISECONDS).
+            atMost(50, TimeUnit.MILLISECONDS).
+        untilAsserted(() -> {
+          assertThat(Request.Get(resultsURL).execute().returnContent().asString()).isNotBlank(); 
+        });
+
+        response = Request.Get(resultsURL).execute().returnContent().asString();
+        assertThat(response).isEqualTo("1\n");
```
Comment (Ladicek) [inline hollow-jar/full/src/test/java/org/wildfly/swarm/ts/hollow/jar/full/FullHollowJarIT.java]: This is the exact assert that should be inside `await().untilAsserted(...)`.

File: pom.xml [hunk 14]
```diff
@@ -112,7 +114,8 @@
         <version.org.jboss.spec.javax.json.jboss-json-api_1.0_spec>1.0.0.Final</version.org.jboss.spec.javax.json.jboss-json-api_1.0_spec>
         <version.org.jboss.spec.javax.servlet.jboss-servlet-api_3.0_spec>1.0.2.Final</version.org.jboss.spec.javax.servlet.jboss-servlet-api_3.0_spec>
         <version.org.jboss.spec.javax.ws.rs.jboss-jaxrs-api_2.0_spec>1.0.0.Final</version.org.jboss.spec.javax.ws.rs.jboss-jaxrs-api_2.0_spec>
-        <version.org.jboss.spec.web.7.0.api>1.1.0.Final</version.org.jboss.spec.web.7.0.api>
+        <version.org.jboss.spec.web.7.0.api>1.1.1.Final</version.org.jboss.spec.web.7.0.api>
```
Comment (Ladicek) [inline pom.xml]: Why did you move back to 1.1.1.Final? See https://github.com/Ladicek/swarm-test-suite/commit/871cdc990f7f63852368c7312c0622712f7959fd

File: pom.xml [hunk 15]
```diff
@@ -328,6 +331,25 @@ remove this line for bom-certified -->
...
+            </dependency>
+            <dependency>
+                <groupId>com.icegreen</groupId>
+                <artifactId>greenmail</artifactId>
+                <version>${version.com.icegreen.greenmail}</version>
+            </dependency>
+            <dependency>
+                 <groupId>net.java.xadisk</groupId>
+                 <artifactId>xadisk</artifactId>
+                 <type>jar</type>
```
Comment (Ladicek) [inline pom.xml]: `<type>jar</type>` is the default, can omit that.

File: pom.xml (hunk 14 above)
Comment (juangon) [reply pom.xml]: Oh f**k I was aware of your change, seems I messed up in the conflict rebase :-(

**Comment by juangon** [comment]
@Ladicek pushed last suggested changes

**Comment by Ladicek** [comment]
Looks good. I'll pull, do some cosmetic changes and merge.
